chapters/lexicalstructure.tex

```diff
@@ -247,6 +247,29 @@ \subsection{Strings}\label{strings}
 \end{lstlisting}
 \end{nonnormative}
 
+\subsection{Units of Literal Constants}\label{units-literal-constants}
+
+The following rules apply:
+\begin{itemize}
+\item The result of \lstinline!x + L!, \lstinline!L + x\lstinline!, \lstinline!x - L!, \lstinline!L - x!, \lstinline!x*L!, \lstinline!L*x!, \lstinline!x/L!, where \lstinline!x! is an expression with non-empty \lstinline!unit! attribute string `<s>` and \lstinline!L! is an expression containing only literal constants, shall have the same \lstinline!unit! attribute string of \lstinline!x!.
```

Comment:
I have a proposal to resolve the distinction between built-in (`exp()`) and other (`specifiEnthalpy()`) functions. Because there's some context to write up around this, see the gist https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea

I also collected all the test cases I could find so far in one place, so we could cross-check against them (and maybe add to a test suite later):  https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea#file-mcp-0027-test-examples-md

I started a new discussion thread because the previous one grew huge and hard to view.

Reply:
@casella any thoughts on this approach to express the distinction between builtin and other functions?

Reply:
@HansOlsson @henrikt-ma @casella it would be great to get your feedback on my comment https://github.com/modelica/ModelicaSpecification/pull/3266/files#r1049500278 with a proposed approach to resolve the distinction between built-in (`exp()`) and other (`specificEnthalpy()`) functions, and hopefully make progress on this PR. :-)

Reply:
I believe that built-in functions should be treated uniformly with other functions as much as possible, and we need to distinguish it in some other way. Specifically specificEnthalphy is a function where the interface-variables have non-empty units, and that should be preserved. Note that some built-in functions have unit (specifically for Angles) in Modelica.Math, which is kind of identical to having `unit="1"`. 

So in one sense many built-in functions can be seen as having `unit="1"` meaning that we can limit literal-part to functions where the interface is without unit. However, having an interface for built-in functions will stumble on `abs`, `sign`, `sqrt`, `atan2`, etc where it makes more sense to say that the unit-handling is defined according to the mathematical rules.

Additionally for deriving units it makes sense that each function call is treated separately; so that if a user defined their own abs-function it is possible to have different units for each call. With such a rule we can just exclude function calls from the literal part; as `x+myabs(2)` will be a two-step procedure: `myabs#1.input=2`, `x+myabs#1.output`.

Reply:
> @HansOlsson @henrikt-ma @casella it would be great to get your feedback on my comment https://github.com/modelica/ModelicaSpecification/pull/3266/files#r1049500278 with a proposed approach to resolve the distinction between built-in (`exp()`) and other (`specificEnthalpy()`) functions, and hopefully make progress on this PR. :-)

@bilderbuchi I apologize. During the last few months I was overwhelmed by teaching load and other stuff, and I couldn't keep track with this discussion. Now, as I get back to it, I would have to restart from the beginning and go through 100+ comments, in multiple branches and sub-branches. I would need a map to navigate this now, otherwise just getting back on track would require to spend several hours to re-read the whole material. To me, this is hopeless, I'm sorry but I can't handle this now.

I think the only way to get through this is to _physically_ lock a few of us in a room ([Viterbo-style](https://en.wikipedia.org/wiki/1268%E2%80%931271_papal_election)), throw out the key and only let us leave when we have a finished proposal 😅. BTW, this is how we made progress with features such as the stream connectors, back in the old days of yore when we didn't have Zoom and GIT and we met in person during 3-days Modelica Design Meetings do do stuff.

If we could somehow draw a line and restart the discussion clean from a point on which we have some general consensus, that would be great. But I'm not sure such a point even exists. In any case, I'm afraid a PR with 100+ comments is dead, or at least not feeling well.

BTW, I'm not even sure how could I actually _read_ the current proposal. Is there a way to get this branch compiled into html or PDF?

Reply:
No need to say sorry, I recognize your points myself. I understand that a conclave-style workshop would be preferable, but I'm not sure if such an approach is in the cards?

However, IIRC the way it seemed to me at least was that this proposal was near ~consensus (or readyness for review), and most conversation threads concluded.  The main remaining item was I think to resolve the precision/scope of the "expression containing only literal constants" formulation, which I attempted again back in December with this Gist.

Reply:
> BTW, I'm not even sure how could I actually read the current proposal. Is there a way to get this branch compiled into html or PDF?

The CI system compiles a PDF, you have to navigate to the build checks below, on the Jenkins page hit the "ARtifacts" button in the upper right, and download the pdf. In this case, this link: https://test.openmodelica.org/jenkins/job/ModelicaAssociation/job/ModelicaSpecification/job/PR-3266/23/artifact/MLS.pdf
Pretty neat I have to say, hats off to the MA dev that set this up! :-)

Reply:
To summarize the situation hopefully more simply (from my point of view/recollection):
* What we _have_ already reached full consensus on right now is the **desired behaviour** of unit checking while avoiding the "wildcard effect", as codified in the set of unit-test-like examples that I collected from the discussions in one place [here](https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea#file-mcp-0027-test-examples-md)
* _What remains_ is to find out the/a **minimal change to the spec** that corresponds to that desired behaviour.
* Your PR here is an attempt to formulate such a change.
* The _one thing_ your PR does **not yet cover correctly** is the _last_ example in the section https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea#proper-functions (as called out in that section). 
* My [proposal update](https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea#file-mcp-0027-proposal-update-md) BB2a **rectifies this deficiency** AFAICT.
* I'm asking if this proposal BB2a is acceptable in content/principle. If yes, I also ask if the refactored formulation BB2b, that streamlines the whole diff, is agreeable. I consider the second point optional if it turns out to be contentious.

Reply:
As it was pointed out below in https://github.com/modelica/ModelicaSpecification/pull/3266#issuecomment-1582197249, the language group is not ready to incorporate a solution to the wildcard effect in isolation.  Instead, it seems that what we have learned in these discussions need to be carried over to a fresh start on unit checking where we try to break down the problem into smaller pieces, and where we in the end will be able to see the complete picture of unit checking before merging anything to the _master_ branch.

Reply:
I missed this recent comment in the notifications/PR discussion.
After a couple of months of mostly productive back-and-forth iteration, and half a year of inactivity, that certainly feels like a disappointing outcome to me. 
One can only hope that, after having failed to achieve sufficient consensus on this minimized-scope problem, making the scope of the problem _larger_ will enable achieving that consensus somehow. This sounds unintuitive to me 😅, but if this is the way this works, hopefully such a holistic perspective will succeed in unblocking things for the language group. 🤞 

I hope that at least the [list of test cases](https://gist.github.com/bilderbuchi/317bc1a2e039f477462a61f5826260ea#file-mcp-0027-test-examples-md) we all came up with together will be useful during these future discussions.